### PR TITLE
Floating interdomain

### PR DESCRIPTION
**Pull request by artem-belov**
## Motivation and Context
#1592

## How Has This Been Tested?

- [ ] Covered by existing integration testing
- [x] Added integration testing to cover
- [ ] Tested locally
- [ ] Have not tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: applications/nsmrs/build/Dockerfile.nsmrs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ENV GOPROXY=${GOPROXY}
+ENV PACKAGEPATH=github.com/networkservicemesh/networkservicemesh/applications/nsmrs
+ENV GO111MODULE=on
+
+RUN mkdir /root/networkservicemesh
+ADD [".","/root/networkservicemesh"]
+WORKDIR /root/networkservicemesh/applications/nsmrs
+RUN VENDORING=${VENDORING} ../../scripts/go-mod-download.sh
+
+RUN CGO_ENABLED=0 GOOS=linux go build ${VENDORING} -ldflags "-extldflags '-static' -X  main.version=${VERSION}" -o /go/bin/nsmrs ./cmd
```
Comment (lobkovilya) [inline applications/nsmrs/build/Dockerfile.nsmrs]: Looks like you already build application [here](https://github.com/networkservicemesh/networkservicemesh/pull/1672/files#diff-c31e810773a4c394261f8d76827aaca3R22) so why do we have to do it again in docker image?

File: applications/nsmrs/pkg/serviceregistryserver/cache.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+type nseRegistryCache struct {
+	networkServiceEndpoints map[string][]*registry.NSERegistration
+	endpoints               map[string]*registry.NSERegistration
+	nseExpirationTimeout    time.Duration
+}
+
+//NewNSERegistryCache creates new nerwork service endpoints cache
+func NewNSERegistryCache() NSERegistryCache {
+	expirationTimeout := NSEExpirationTimeoutDefault
+	if interval := strings.TrimSpace(os.Getenv(NSEExpirationTimeoutSecondsEnv)); interval != "" {
```
Comment (lobkovilya) [inline applications/nsmrs/pkg/serviceregistryserver/cache.go]: Please use `utils.EnvVar` from `utils/env.go` to read and parse envs

File: applications/nsmrs/pkg/serviceregistryserver/cache.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	return &nseRegistryCache{
+		networkServiceEndpoints: make(map[string][]*registry.NSERegistration),
+		endpoints:               make(map[string]*registry.NSERegistration),
+		nseExpirationTimeout:    expirationTimeout,
+	}
+}
+
+// AddNetworkServiceEndpoint - register NSE in cache
+func (rc *nseRegistryCache) AddNetworkServiceEndpoint(entry *registry.NSERegistration) (*registry.NSERegistration, error) {
+	if endpoint, ok := rc.endpoints[entry.NetworkServiceEndpoint.Name]; ok {
```
Comment (lobkovilya) [inline applications/nsmrs/pkg/serviceregistryserver/cache.go]: `rc.endpoints` is modified from different goroutines, please use some primitives of synchronization

File: applications/build.mk [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+# limitations under the License.
+
+apps = $(shell ls -d ./applications/*/ | xargs -n 1 basename)
+apps_targets = $(addsuffix -build, $(addprefix go-, $(apps)))
+
+.PHONY: $(apps_targets)
+$(apps_targets): go-%-build:
+	@echo "----------------------  Building applications::$* via Cross compile ----------------------" && \
+	pushd ./applications/$* && \
+	CGO_ENABLED=$(CGO_ENABLED) GOOS=$(GOOS) go build \
```
Comment (haiodo) [inline applications/build.mk]: Did we already have this kind of target in some defines? I would prefer for us to have all compile options to be in one place.

File: applications/nsmrs/pkg/serviceregistryserver/cache.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	// NSEExpirationTimeoutEnv - environment variable contains custom NSEExpirationTimeout
+	NSEExpirationTimeoutEnv = utils.EnvVar("NSE_EXPIRATION_TIMEOUT")
+)
+
+// NSERegistryCache - cache of registered Network Service Endpoints
+type NSERegistryCache interface {
+	AddNetworkServiceEndpoint(nse *registry.NSERegistration) (*registry.NSERegistration, error)
+	UpdateNetworkServiceEndpoint(nse *registry.NSERegistration) (*registry.NSERegistration, error)
+	DeleteNetworkServiceEndpoint(endpointName string) (*registry.NSERegistration, error)
+	GetEndpointsByNs(networkServiceName string) []*registry.NSERegistration
```
Comment (haiodo) [inline applications/nsmrs/pkg/serviceregistryserver/cache.go]: May be name this method as GetEndpoints? Why we need ByNs if we have just one method.

File: applications/nsmrs/pkg/serviceregistryserver/cache.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+	StartNSMDTracking()
```
Comment (haiodo) [inline applications/nsmrs/pkg/serviceregistryserver/cache.go]: Name is not obvious why it should be in interface.

File: applications/nsmrs/pkg/serviceregistryserver/cache.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+// GetEndpointsByNs - get Endpoints list from cache by Name
+func (rc *nseRegistryCache) GetEndpointsByNs(networkServiceName string) []*registry.NSERegistration {
+	return rc.networkServiceEndpoints[networkServiceName]
+}
+
+func (rc *nseRegistryCache) StartNSMDTracking() {
+	go func() {
+		for {
+			<-time.After(rc.nseExpirationTimeout / 2)
```
Comment (haiodo) [inline applications/nsmrs/pkg/serviceregistryserver/cache.go]: I think we need to wait for minimal NSE lifetime left, and react on add/update/remove of NSEs. I do not think we need to wakeup so often.

File: applications/nsmrs/pkg/serviceregistryserver/registry.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+// NewNseRegistryService - creates NSE Registry service
+func NewNseRegistryService(cache NSERegistryCache) NSERegistryService {
+	return &nseRegistryService{
+		cache: cache,
+	}
+}
+
+// RegisterNSE - Registers NSE in cache
+func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry.NSERegistration) (*registry.NSERegistration, error) {
+	logrus.Infof("Received RegisterNSE(%v)", request)
```
Comment (haiodo) [inline applications/nsmrs/pkg/serviceregistryserver/registry.go]: Please use spanhelper to send spans.

File: applications/nsmrs/pkg/serviceregistryserver/registry.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+		logrus.Errorf("Error registering NSE: %v", err)
+		return nil, err
+	}
+
+	logrus.Infof("Returned from RegisterNSE: request: %v", request)
+	return request, err
+}
+
+func (rs *nseRegistryService) BulkRegisterNSE(srv registry.NetworkServiceRegistry_BulkRegisterNSEServer) error {
+	for {
```
Comment (haiodo) [inline applications/nsmrs/pkg/serviceregistryserver/registry.go]: Please use spanhelper to send spans.

File: applications/nsmrs/pkg/serviceregistryserver/registry.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+		if err != nil {
+			err = errors.Errorf("error processing BulkRegisterNSE request: %v", err)
+			return err
+		}
+	}
+}
+
+// RemoveNSE - Removes NSE from cache and stops NSMgr monitor
+func (rs *nseRegistryService) RemoveNSE(ctx context.Context, request *registry.RemoveNSERequest) (*empty.Empty, error) {
+	logrus.Infof("Received RemoveNSE(%v)", request)
```
Comment (haiodo) [inline applications/nsmrs/pkg/serviceregistryserver/registry.go]: Please use spanhelper to send spans.

File: applications/nsmrs/pkg/serviceregistryserver/server.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+	return net.Listen("tcp", registryAPIAddress)
+}
+
+// NewNSMDServiceRegistryServer - creates new service registry service
+func NewNSMDServiceRegistryServer() ServiceRegistry {
+	return &serviceRegistry{}
+}
+
+// New - creates new grcp server and registers NSE discovery and registry services
+func New() *grpc.Server {
```
Comment (haiodo) [inline applications/nsmrs/pkg/serviceregistryserver/server.go]: Better to pass context here and report spans.

File: deployments/helm/nsmrs/templates/nsmrs.tpl [hunk 12]
```diff
@@ -0,0 +1,39 @@
+---
+apiVersion: apps/v1
+kind: DaemonSet
```
Comment (haiodo) [inline deployments/helm/nsmrs/templates/nsmrs.tpl]: Do we really need it on every node? Is it intended?

File: k8s/pkg/proxyregistryserver/registry.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+func newNseRegistryService(clusterInfoService clusterinfo.ClusterInfoServer) *nseRegistryService {
+	return &nseRegistryService{
+		clusterInfoService: clusterInfoService,
+	}
+}
+
+func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry.NSERegistration) (*registry.NSERegistration, error) {
+	logrus.Infof("%s: received RegisterNSE(%v)", NSRegistryForwarderLogPrefix, request)
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: Please use spanhelper to send spans.

File: k8s/pkg/proxyregistryserver/registry.go [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+	// Swapping IP address to external (keep port)
+	url := request.NetworkServiceManager.Url
+	if idx := strings.Index(url, ":"); idx > -1 {
+		externalIP += url[idx:]
+	}
+	request.NetworkServiceManager.Url = externalIP
+
+	logrus.Infof("%s: Prepared forwarding RegisterNSE request: %v", NSRegistryForwarderLogPrefix, request)
+
+	remoteRegistry := nsmd.NewServiceRegistryAt(nsmrsURL + ":80")
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: Should port 80 should be configurable?

File: k8s/pkg/proxyregistryserver/registry.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+		errIn := errors.Errorf("failed register NSE in NSMRS: %v", err)
+		logrus.Errorf("%s: %v", NSRegistryForwarderLogPrefix, errIn)
+		return request, errIn
+	}
+
+	return request, nil
+}
+
+func (rs *nseRegistryService) BulkRegisterNSE(srv registry.NetworkServiceRegistry_BulkRegisterNSEServer) error {
+	logrus.Infof("%s: Forwarding Bulk Register NSE stream...", NSRegistryForwarderLogPrefix)
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: Please use spanhelper to send spans.

File: k8s/pkg/proxyregistryserver/registry.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+				break
+			}
+		}
+
+		<-time.After(NSMRSReconnectInterval)
+	}
+}
+
+func requestBulkRegisterNSEStream(ctx context.Context, remoteRegistry serviceregistry.ServiceRegistry, nsmrsURL string) (registry.NetworkServiceRegistry_BulkRegisterNSEClient, error) {
+	nseRegistryClient, err := remoteRegistry.NseRegistryClient(ctx)
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: Please use spanhelper to send spans.

File: k8s/pkg/proxyregistryserver/registry.go [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+	stream, err := nseRegistryClient.BulkRegisterNSE(ctx)
+	if err != nil {
+		return nil, errors.Errorf("error forwarding BulkRegisterNSE request to %s : %v", nsmrsURL, err)
+	}
+
+	return stream, nil
+}
+
+func (rs *nseRegistryService) RemoveNSE(ctx context.Context, request *registry.RemoveNSERequest) (*empty.Empty, error) {
+	logrus.Infof("%s: Received RemoveNSE(%v)", NSRegistryForwarderLogPrefix, request)
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: Please use spanhelper to send spans.

File: k8s/pkg/registryserver/nse.go [hunk 18]
```diff
@@ -89,11 +99,74 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+				logrus.Errorf("Cannot forward NSE Registration: %v", forwardErr)
+			}
+		}()
 	}
 	logrus.Infof("Returned from RegisterNSE: time: %v request: %v", time.Since(st), request)
 	return request, nil
 }
 
+func (rs *nseRegistryService) BulkRegisterNSE(srv registry.NetworkServiceRegistry_BulkRegisterNSEServer) error {
+	logrus.Infof("Forwarding Bulk Register NSE stream...")
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use spanhelper to send spans.

File: k8s/pkg/registryserver/nse.go [hunk 19]
```diff
@@ -89,11 +99,74 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+				break
+			}
+		}
+
+		<-time.After(ProxyRegistryReconnectInterval)
+	}
+}
+
+func requestBulkRegisterNSEStream(ctx context.Context, remoteRegistry serviceregistry.ServiceRegistry, nsrURL string) (registry.NetworkServiceRegistry_BulkRegisterNSEClient, error) {
+	nseRegistryClient, err := remoteRegistry.NseRegistryClient(ctx)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use spanhelper to send spans.

File: k8s/pkg/registryserver/nse.go [hunk 20]
```diff
@@ -102,6 +175,128 @@ func (rs *nseRegistryService) RemoveNSE(ctx context.Context, request *registry.R
...
+			logrus.Errorf("Cannot forward Remove NSE: %v", forwardErr)
+		}
+	}()
+
 	logrus.Infof("RemoveNSE done: time %v", time.Since(st))
 	return &empty.Empty{}, nil
 }
+
+func (rs *nseRegistryService) forwardRegisterNSE(request *registry.NSERegistration) error {
+	logrus.Infof("Forwarding Register NSE request (%v)", request)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use spanhelper to send spans.

File: k8s/pkg/registryserver/nse.go [hunk 21]
```diff
@@ -102,6 +175,128 @@ func (rs *nseRegistryService) RemoveNSE(ctx context.Context, request *registry.R
...
+	_, err = nseRegistryClient.RegisterNSE(ctx, request)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (rs *nseRegistryService) forwardRemoveNSE(request *registry.RemoveNSERequest) error {
+	logrus.Infof("Forwarding Remove NSE request (%v)", request)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use spanhelper to send spans.

File: applications/nsmrs/cmd/main.go [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+		logrus.Errorf("Failed to start Public API server...")
+		return
+	}
+
+	startAPIServerAt(sock)
+
+	<-c
+}
+
+func startAPIServerAt(sock net.Listener) {
```
Comment (denis-tingajkin) [inline applications/nsmrs/cmd/main.go]: Should we add a probe for this service?

File: applications/nsmrs/pkg/serviceregistryserver/cache.go [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+
+const (
+	// NSEExpirationTimeoutDefault - default Endpoint expiration timeout, NSE will be deleted if UpdateNetworkServiceEndpoint not received
+	NSEExpirationTimeoutDefault = 5 * time.Minute
+	// NSEExpirationTimeoutEnv - environment variable contains custom NSEExpirationTimeout
+	NSEExpirationTimeoutEnv = utils.EnvVar("NSE_EXPIRATION_TIMEOUT")
+)
+
+// NSERegistryCache - cache of registered Network Service Endpoints
+type NSERegistryCache interface {
```
Comment (denis-tingajkin) [inline applications/nsmrs/pkg/serviceregistryserver/cache.go]: Should we add context.Context in all methods of this interface to add spans?

File: controlplane/pkg/nsmd/registry.go [hunk 24]
```diff
@@ -134,3 +155,38 @@ func (es *registryServer) RemoveNSE(ctx context.Context, request *registry.Remov
...
+	}
+
+	trackingInterval := NSETrackingIntervalSecondsEnv.GetOrDefaultDuration(NSETrackingIntervalDefault)
+	go func() {
+		defer cancel()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
```
Comment (denis-tingajkin) [inline controlplane/pkg/nsmd/registry.go]: let's add log something for this branch

File: k8s/pkg/proxyregistryserver/registry.go [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+	if strings.TrimSpace(nsmrsURL) == "" {
+		err := errors.Errorf("NSMRS Address variable was not set")
+		logrus.Warnf("%s: Skipping Bulk Register NSE forwarding: %v", NSRegistryForwarderLogPrefix, err)
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(srv.Context())
+	defer cancel()
+
+	remoteRegistry := nsmd.NewServiceRegistryAt(nsmrsURL + ":80")
```
Comment (denis-tingajkin) [inline k8s/pkg/proxyregistryserver/registry.go]: Move `80` can be constant

File: k8s/pkg/proxyregistryserver/registry.go [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+
+func requestBulkRegisterNSEStream(ctx context.Context, remoteRegistry serviceregistry.ServiceRegistry, nsmrsURL string) (registry.NetworkServiceRegistry_BulkRegisterNSEClient, error) {
+	nseRegistryClient, err := remoteRegistry.NseRegistryClient(ctx)
+	if err != nil {
+		return nil, errors.Errorf("error forwarding BulkRegisterNSE request to %s : %v", nsmrsURL, err)
+	}
+
+	stream, err := nseRegistryClient.BulkRegisterNSE(ctx)
+	if err != nil {
+		return nil, errors.Errorf("error forwarding BulkRegisterNSE request to %s : %v", nsmrsURL, err)
```
Comment (denis-tingajkin) [inline k8s/pkg/proxyregistryserver/registry.go]: `errors.Wrapf`

File: k8s/pkg/registryserver/nse.go [hunk 27]
```diff
@@ -102,6 +175,128 @@ func (rs *nseRegistryService) RemoveNSE(ctx context.Context, request *registry.R
...
+			return
+		}
+
+		done <- nseRegistryClient
+	}()
+
+	var nseRegistryClient registry.NetworkServiceRegistryClient
+	select {
+	case nseRegistryClient = <-done:
+		// continue
```
Comment (denis-tingajkin) [inline k8s/pkg/registryserver/nse.go]: use `break` here

File: k8s/pkg/registryserver/nse.go [hunk 28]
```diff
@@ -102,6 +175,128 @@ func (rs *nseRegistryService) RemoveNSE(ctx context.Context, request *registry.R
...
```
Comment (denis-tingajkin) [inline k8s/pkg/registryserver/nse.go]: `break`

File: k8s/pkg/apis/networkservice/v1alpha1/zz_generated.deepcopy.go [hunk 29]
```diff
@@ -2,7 +2,7 @@
 
 // Copyright (c) 2019 Cisco and/or its affiliates.
 // Copyright (c) 2019 Red Hat Inc. and/or its affiliates.
-// Copyright (c) 2019 VMware, Inc.
+// Copyright (c) 2019 VMware, Inc
```
[the remark this answers is not in the document]
Comment (artem-belov) [reply k8s/pkg/apis/networkservice/v1alpha1/zz_generated.deepcopy.go]: It is automatically generated file

File: controlplane/pkg/nsmd/registry.go [hunk 30]
```diff
@@ -103,9 +115,19 @@ func (es *registryServer) RegisterNSEWithClient(ctx context.Context, request *re
...
+	if err != nil {
+		logrus.Infof("Error starting NSE tracking requests : %v", err)
+	}
+
 	return registration, nil
 }
 
+func (es *registryServer) BulkRegisterNSE(srv registry.NetworkServiceRegistry_BulkRegisterNSEServer) error {
+	<-srv.Context().Done()
+	return nil
```
[the remark this answers is not in the document]
Comment (artem-belov) [reply controlplane/pkg/nsmd/registry.go]: BulkRegisterNSE at this point does not involved in NSE update process. So it is not necessary

File: applications/nsmrs/cmd/main.go (hunk 22 above)
Comment (artem-belov) [reply applications/nsmrs/cmd/main.go]: I do not think so. NSMRS server is designed to be used out of kubernetes

File: applications/nsmrs/pkg/serviceregistryserver/cache.go (hunk 23 above)
Comment (artem-belov) [reply applications/nsmrs/pkg/serviceregistryserver/cache.go]: I am not sure that it is necessary. There is no important logs in Cache functions

File: applications/nsmrs/pkg/serviceregistryserver/cache.go (hunk 7 above)
Comment (artem-belov) [reply applications/nsmrs/pkg/serviceregistryserver/cache.go]: If we will change interval to next NSE timeout each request - it will go through whole NSE list much more often then now

File: applications/nsmrs/cmd/main.go [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+package main
+
+import (
+	"context"
+	"math/rand"
+	"net"
+	"time"
+
+	"github.com/networkservicemesh/networkservicemesh/pkg/tools/spanhelper"
+
```
Comment (alex-yust) [inline applications/nsmrs/cmd/main.go]: remove blanks?

File: applications/nsmrs/cmd/main.go [hunk 32]
```diff
@@ -0,0 +1,86 @@
...
+	span.Logger().Println("Starting NSMD Service Registry Server on " + address)
+	serviceRegistryServer := serviceregistryserver.NewNSMDServiceRegistryServer()
+	sock, err := serviceRegistryServer.NewPublicListener(address)
+	if err != nil {
+		span.Logger().Errorf("Failed to start Public API server...")
+		return
+	}
+
+	startAPIServerAt(span.Context(), sock)
+
```
Comment (haiodo) [inline applications/nsmrs/cmd/main.go]: Please do span.Finish() here, since start is complete at this point.

File: k8s/pkg/proxyregistryserver/registry.go [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+	logger.Infof("%s: received RegisterNSE(%v)", NSRegistryForwarderLogPrefix, request)
+
+	nsmrsURL := os.Getenv(NSMRSAddressEnv)
+	if strings.TrimSpace(nsmrsURL) == "" {
+		err := errors.Errorf("NSMRS Address variable was not set")
+		logger.Warnf("%s: Skipping Register NSE forwarding: %v", NSRegistryForwarderLogPrefix, err)
+		return request, err
+	}
+
+	nodeConfiguration, cErr := rs.clusterInfoService.GetNodeIPConfiguration(ctx, &clusterinfo.NodeIPConfiguration{NodeName: request.NetworkServiceManager.Name})
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: Please use span.Context() to pass it to child method to properly build a tree of spans.

File: k8s/pkg/proxyregistryserver/registry.go [hunk 34]
```diff
@@ -0,0 +1,203 @@
...
+		externalIP += url[idx:]
+	}
+	request.NetworkServiceManager.Url = externalIP
+
+	logger.Infof("%s: Prepared forwarding RegisterNSE request: %v", NSRegistryForwarderLogPrefix, request)
+
+	remoteRegistry := nsmd.NewServiceRegistryAt(nsmrsURL)
+	defer remoteRegistry.Stop()
+
+	nseRegistryClient, err := remoteRegistry.NseRegistryClient(context.Background())
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: span.Context()

File: k8s/pkg/proxyregistryserver/registry.go [hunk 35]
```diff
@@ -0,0 +1,203 @@
...
+	if strings.TrimSpace(nsmrsURL) == "" {
+		err := errors.Errorf("NSMRS Address variable was not set")
+		logger.Warnf("%s: Skipping Register NSE forwarding: %v", NSRegistryForwarderLogPrefix, err)
+		return &empty.Empty{}, err
+	}
+
+	remoteRegistry := nsmd.NewServiceRegistryAt(nsmrsURL)
+	defer remoteRegistry.Stop()
+
+	nseRegistryClient, err := remoteRegistry.NseRegistryClient(context.Background())
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: Please use span.Context() instead of ctx

File: k8s/pkg/registryserver/nse.go [hunk 36]
```diff
@@ -89,19 +104,220 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+
+	logger.Infof("Received RemoveNSE(%v)", request)
 
 	if err := rs.cache.DeleteNetworkServiceEndpoint(request.GetNetworkServiceEndpointName()); err != nil {
 		return nil, err
 	}
-	logrus.Infof("RemoveNSE done: time %v", time.Since(st))
+
+	go func() {
+		if forwardErr := rs.forwardRemoveNSE(ctx, request); forwardErr != nil {
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use span.Context() instead of ctx

File: k8s/pkg/registryserver/nse.go [hunk 37]
```diff
@@ -89,19 +104,220 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+	logger := span.Logger()
+
+	logger.Infof("Forwarding Register NSE request (%v)", request)
+
+	nsrURL := os.Getenv(ProxyNsmdK8sAddressEnv)
+	if strings.TrimSpace(nsrURL) == "" {
+		nsrURL = ProxyNsmdK8sAddressDefaults
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), ForwardingTimeout)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use spanhelper.CopySpan() with context.Background() if you would like to have new context, or use span.Context() if same is ok.

File: k8s/pkg/registryserver/nse.go [hunk 38]
```diff
@@ -89,19 +104,220 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+	defer cancel()
+
+	done := make(chan registry.NetworkServiceRegistryClient)
+	quit := make(chan error)
+
+	remoteRegistry := nsmd.NewServiceRegistryAt(nsrURL)
+	defer remoteRegistry.Stop()
+
+	go func() {
+		nseRegistryClient, err := remoteRegistry.NseRegistryClient(ctx)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use span.Context() instead of ctx

File: k8s/pkg/registryserver/nse.go [hunk 39]
```diff
@@ -89,19 +104,220 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+		}
+
+		match := &registry.Match{
+			SourceSelector: m.SourceSelector,
+			Routes:         routes,
+		}
+		request.NetworkService.Matches = append(request.NetworkService.Matches, match)
+	}
+
+	_, err = nseRegistryClient.RegisterNSE(ctx, request)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use span.Context() instead of ctx

File: k8s/pkg/registryserver/nse.go [hunk 40]
```diff
@@ -89,19 +104,220 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+	logger := span.Logger()
+
+	logger.Infof("Forwarding Remove NSE request (%v)", request)
+
+	nsrURL := os.Getenv(ProxyNsmdK8sAddressEnv)
+	if strings.TrimSpace(nsrURL) == "" {
+		nsrURL = ProxyNsmdK8sAddressDefaults
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), ForwardingTimeout)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Same as on Register.

File: k8s/pkg/registryserver/nse.go [hunk 41]
```diff
@@ -89,19 +104,220 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+	defer cancel()
+
+	done := make(chan registry.NetworkServiceRegistryClient)
+	quit := make(chan error)
+
+	remoteRegistry := nsmd.NewServiceRegistryAt(nsrURL)
+	defer remoteRegistry.Stop()
+
+	go func() {
+		nseRegistryClient, err := remoteRegistry.NseRegistryClient(ctx)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use span.Context() instead of ctx

File: k8s/pkg/registryserver/nse.go [hunk 42]
```diff
@@ -89,19 +104,220 @@ func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry
...
+	select {
+	case nseRegistryClient = <-done:
+		break
+	case err := <-quit:
+		return err
+	case <-time.After(ForwardingTimeout):
+		return errors.Errorf("timeout requesting NseRegistryClient")
+	}
+
+	_, err := nseRegistryClient.RemoveNSE(ctx, request)
```
Comment (haiodo) [inline k8s/pkg/registryserver/nse.go]: Please use span.Context() instead of ctx

File: k8s/pkg/registryserver/server.go [hunk 43]
```diff
@@ -40,9 +39,9 @@ func New(ctx context.Context, clientset *nsmClientset.Clientset, nsmName string)
 	registry.RegisterNsmRegistryServer(server, nsmRegistry)
 
 	if err := cache.Start(); err != nil {
-		logrus.Error(err)
+		span.Logger().Error(err)
```
Comment (haiodo) [inline k8s/pkg/registryserver/server.go]: Please use span.LogError(err) it will check err for nil and add it to span.

File: k8s/pkg/proxyregistryserver/registry.go [hunk 44]
```diff
@@ -0,0 +1,203 @@
...
+	"time"
+
+	"github.com/networkservicemesh/networkservicemesh/pkg/tools/spanhelper"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/serviceregistry"
+
+	"github.com/pkg/errors"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"golang.org/x/net/context"
```
Comment (haiodo) [inline k8s/pkg/proxyregistryserver/registry.go]: Should it be ""context" instead?

File: docs/spec/interdomain.md [hunk 45]
```diff
@@ -28,6 +28,19 @@ Interdomain NSM does not have central registry. All clusters are communicate jus
...
+Floating interdomain provides an ability to register Network Service Endpoints from any domain at one place, available from any another domain.
+
+* NSMRS (Network Service Mesh Registry Server) is used as interdomain NSE registry server. 
+* Floating Interdomain handle Network Services requests the same way as regular Interdomain request (for example request for Network Service of the form *network-service@nsmrs-domain.com*)
+* NSMRS is independent from kubernetes (except spire registration).
+* Proxy NSMD-K8S should be configured to forward registry packets to the NSMRS (Environment variable "*NSMRS_ADDRESS*").
+* Helm script require specific label for node to assign NSMRS deployment
+> kubectl label nodes \<node-name\> nsmrs=true
+
+
```
Comment (nickolaev) [inline docs/spec/interdomain.md]: ~2000 lines of code and just 7 lines of doc?

File: forwarder/build.mk [hunk 46]
```diff
@@ -18,16 +18,14 @@
 go-kernel-forwarder-build: go-%-build:
 	$(info ----------------------  Building forwarder::$* via Cross compile ----------------------)
 	@pushd ./forwarder && \
-	CGO_ENABLED=$(CGO_ENABLED) GOOS=$(GOOS) go build \
-    	-ldflags "-extldflags '-static' -X  main.version=$(VERSION)" -o $(BIN_DIR)/$*/$* ./kernel-forwarder/cmd/ && \
+	${GO_BUILD} -o $(BIN_DIR)/$*/$* ./kernel-forwarder/cmd/ && \
```
Comment (nickolaev) [inline forwarder/build.mk]: Please move this into its own PR.

File: test/build.mk [hunk 47]
```diff
@@ -21,8 +21,7 @@ test_targets = $(addsuffix -build, $(addprefix go-, $(test_apps)))
 $(test_targets): go-%-build:
 	@echo "----------------------  Building test/applications::$* via Cross compile ----------------------" && \
 	pushd ./test && \
-	CGO_ENABLED=$(CGO_ENABLED) GOOS=$(GOOS) go build \
-    	-ldflags "-extldflags '-static' -X  main.version=$(VERSION)" -o $(BIN_DIR)/$*/$* ./applications/cmd/$* && \
+	${GO_BUILD} -o $(BIN_DIR)/$*/$* ./applications/cmd/$* && \
```
Comment (nickolaev) [inline test/build.mk]: Separete PR please.

File: deployments/helm/nsmrs/Chart.yaml [hunk 48]
```diff
@@ -0,0 +1,6 @@
+---
+apiVersion: v1
+appVersion: "1.0"
+description: A Helm chart for Kubernetes
+name: nsmrs
+version: 0.1.0
```
Comment (nickolaev) [inline deployments/helm/nsmrs/Chart.yaml]: versions should reflect the release we have so it's '0.2.0'

File: deployments/helm/nsmrs/Chart.yaml [hunk 49]
```diff
@@ -0,0 +1,6 @@
+---
+apiVersion: v1
+appVersion: "1.0"
```
Comment (nickolaev) [inline deployments/helm/nsmrs/Chart.yaml]: `0.2.0`
